crates/red_knot_python_semantic/src/types/builder.rs

```diff
@@ -284,10 +284,16 @@ impl<'db> InnerIntersectionBuilder<'db> {
                 }
             }
             _ => {
-                let addition_is_bool_instance = new_positive
+                let known_instance = new_positive
                     .into_instance()
-                    .and_then(|instance| instance.class.known(db))
-                    .is_some_and(KnownClass::is_bool);
+                    .and_then(|instance| instance.class.known(db));
+
+                let addition_is_bool_instance = match known_instance {
+                    // `object & T` -> `T`
+                    Some(KnownClass::Object) => return,
+                    Some(KnownClass::Bool) => true,
+                    _ => false,
+                };
```

Comment:
I think I would prefer to split that into two parts for readability(?)
```suggestion
                if known_instance == Some(KnownClass::Object) {
                    // `object & T` -> `T`
                    return;
                }

                let addition_is_bool_instance = known_instance == Some(KnownClass::Bool);
```

Reply:
I wondered about that... I assumed the single `match` would be better for performance, but maybe I'm prematurely optimising (or underestimating the compiler's ability to optimise this code)